astropy/wcs/wcsapi/fitswcs.py

```diff
@@ -401,7 +401,12 @@ def _get_components_and_classes(self):
 
                 kwargs = {}
                 kwargs['frame'] = celestial_frame
-                kwargs['unit'] = u.deg
+                # Very occasionally (TAB) wcs does not convert the units
+                # We call set here to make sure any unit conversion which is
+                # going to happen has been done.
+                self.wcs.set()
```

Comment:
I am not sure if this is really needed. Maybe @mcara knows?

Reply:
Not sure I understand the issue (i.e., what/when units do not get converted). As to when `set()` needs to be called, basically never: if you already have a valid WCS object and set parameters in a "normal" way like using assignments like:
```python
w.wcs.crval = [20, 30]
```
instead of
```python
w.wcs.ctype[0] = 20
w.wcs.ctype[1] = 30
w.wcs.set()
```
then WCS code will automatically call `set()` as needed. When you do weirder things like second approach above, the yes, you need to call `WCS.Wcsprm.set()`. A full list of when `set` should be called can be found here: https://www.atnf.csiro.au/people/mcalabre/WCS/wcslib/structwcsprm.html#a35bff8de85e5a8892e1b68db69ca7a68

This is a "mixin" class and I worry about your call to `set` interacting with https://github.com/astropy/astropy/blob/b6352e194ab00de75f33f0b36ccb944333d20a10/astropy/wcs/wcs.py#L548-L549

Reply:
The issue I have seen before which I am kind of blindly solving here is that until you trigger a call to `set()` (i.e. implicitly in pixel to world or whatever) `w.wcs.cunit` can be in units of not-degrees however after `set()` has been called they will be in degrees. Before I use the unit here I just wanted to make sure they had been changed if they were going to be changed.

I guess that by the time you get to here in this mixin you have already called `pixel_to_world` or whatever so I guess that this is not needed.

P.S. Does calling `set()` twice cause any issues?

Reply:
calling `set` after it was already called will cause performance issues. It is an expensive operation. You may not feel it if it is done once during the initialization but if this is done repeatedly it will be felt. So it depends on the usage.

My main concern though was that it may call `set` even when `_do_set` is `False`. I do not recall right now why `_do_set` was introduced but there was a purpuse. Let me investigate...

Reply:
Another reason to avoid calling `set` twice is that if there are warning messages, they may also appear twice (but I did not check whether is is indeed the case).

Reply:
I have removed it. I don't think that it's really needed, but I could be wrong